### PR TITLE
606 - Get AP Capacity endpoint

### PR DESCRIPTION
**Pull request by brownbearsoftware**
Add endpoint to get AP capacity per day, from today until the furthest-in-the-future Booking or Lost Beds

### CODE REVIEW

**Comment by brownbearsoftware** [comment]
[Write API endpoint to get the capacity of an AP](https://trello.com/c/vHBidefy/606-write-api-endpoint-to-get-the-capacity-of-an-ap)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/Availability.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import java.time.LocalDate
+
+data class Availability(
+  val date: LocalDate,
+  val pendingBookings: Int,
+  val arrivedBookings: Int,
+  val nonArrivedBookings: Int,
+  val cancelledBookings: Int,
+  val lostBeds: Int
+)
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/Availability.kt]: I think we just said to return the `availableBeds` here? I think this info might be interesting though, so maybe return the `availableBeds` in addition to the count?

Comment (pezholio) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/Availability.kt]: Oh wait, I'm getting ahed of myself here - so to be clear, this is a data class that has all the different types of bookings, which we then sum when calculating returning in the endpoint?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt [hunk 2]
```diff
@@ -150,4 +152,31 @@ class PremisesController(
...
+    val lastLostBedsDate = premisesService.getLastLostBedsDate(premises)
+
+    val capacityForPeriod = premisesService.getAvailabilityForRange(
+      premises,
+      LocalDate.now(),
+      maxOf(
+        LocalDate.now(),
+        lastBookingDate ?: LocalDate.now(),
+        lastLostBedsDate ?: LocalDate.now()
+      )
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt:170]: Love it!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/Availability.kt [hunk 3]
```diff
@@ -9,4 +9,6 @@ data class Availability(
   val nonArrivedBookings: Int,
   val cancelledBookings: Int,
   val lostBeds: Int
-)
+) {
+  fun getFreeCapacity(totalBeds: Int) = totalBeds - pendingBookings - arrivedBookings - lostBeds
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/Availability.kt:13]: Is it worth bracketing this to be clearer?

```suggestion
  fun getFreeCapacity(totalBeds: Int) = ((totalBeds - pendingBookings) - arrivedBookings) - lostBeds
```

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/CapacityTest.kt [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+      .header("Authorization", "Bearer $jwt")
+      .exchange()
+      .expectStatus()
+      .isOk
+      .expectBodyList<DateCapacity>()
+      .hasSize(0)
+  }
+
+  @Test
+  fun `Get Capacity with booking in future returns OK with list entry for each day until the booking ends`() {
```
Comment (pezholio) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/CapacityTest.kt:105]: I dunno if it'd be easier and quicker to break all these edge cases into unit tests, which would be cheaper / faster than having integration tests for all the possible cases. What do you think?

**Review by pezholio — COMMENTED** [review]
Looks good from my POV. Would like some Kotlin folks to review it though. Might also be good to unit test `getAvailabilityForRange` to prevent having a load of integration tests.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/Availability.kt (hunk 1 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/Availability.kt]:
```
Oh wait, I'm getting ahed of myself here - so to be clear, this is a data class that has all the different types of bookings, which we then sum when calculating returning in the endpoint?
```

that's right yeah, I suspect we'll need these counts elsewhere so figured the service layer should return them all then just transform it for the endpoint's response

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt (hunk 2 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt:170]: elvis operator up there with the best kotlin features :D

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/Availability.kt (hunk 3 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/Availability.kt]: Have done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/CapacityTest.kt (hunk 4 above)
Comment (brownbearsoftware) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/CapacityTest.kt:90]: Yeah agreed, I've removed some of the integration tests and added some unit ones

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt [hunk 5]
```diff
@@ -150,4 +152,31 @@ class PremisesController(
 
     return ResponseEntity.ok(lostBedsTransformer.transformJpaToApi(lostBed))
   }
+
+  override fun premisesPremisesIdCapacityGet(premisesId: UUID): ResponseEntity<List<DateCapacity>> {
+    val premises = premisesService.getPremises(premisesId)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt:157]: very very minor - I wonder if there is any point in letting the exception handler process these by throwing, since you are having to deal with ResponseEntity here anyway you could just do something like `return ResponseEntity.notFound(some error )`

Obviously you would have to restructure this sine you want to return early etc etc

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt [hunk 6]
```diff
@@ -13,7 +14,13 @@ import javax.persistence.OneToOne
 import javax.persistence.Table
 
 @Repository
-interface BookingRepository : JpaRepository<BookingEntity, UUID>
+interface BookingRepository : JpaRepository<BookingEntity, UUID> {
+  @Query("SELECT b FROM BookingEntity b WHERE b.premises.id = :premisesId AND b.arrivalDate <= :endDate AND b.departureDate >= :startDate")
+  fun findAllByPremisesIdAndOverlappingDate(premisesId: UUID, startDate: LocalDate, endDate: LocalDate): MutableList<BookingEntity>
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt:19]: just curious, why does this have to return a `MutableList` ?

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt (hunk 5 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt:157]: So we could do but because we're returning an https://datatracker.ietf.org/doc/html/rfc7807 problem+json response we'd end up with quite a lot of duplication everywhere constructing the response bodies

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt (hunk 6 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt]: Would seem it doesn't, have switched to List

**Review by andymarke — APPROVED** [review]
lgtm
